### ISSUE TITLE
TypeVar constraints fail to distribute in unions of generic types

### ISSUE THREAD

**Issue by robert-royals** [microsoft/pyright#11203]
When a `TypeVar` has constraints, it is supposed to mean that the type variable must take the exact value of one of them. For the examples below, this means that for every valid type `T` might take, the typing assignments are sound:

```python
def f[T: (int, str)](t: T) -> tuple[int] | tuple[str]:
    return (t,)


class A[T: (int, str)]:
    t: T
    x: tuple[int] | tuple[str]

    def f(self) -> None:
        self.x = (self.t,)
```
Yet  for `f`, pyright reports:

Type "tuple[T@f]" is not assignable to return type "tuple[int] | tuple[str]"
    Type "tuple[T@f]" is not assignable to type "tuple[int] | tuple[str]"
      "tuple[T@f]" is not assignable to "tuple[int]"
        Tuple entry 1 is incorrect type
          Type "int* | str*" is not assignable to type "int"
            "str*" is not assignable to "int"
      "tuple[T@f]" is not assignable to "tuple[str]"
        Tuple entry 1 is incorrect type
          Type "int* | str*" is not assignable to type "str"
    ... (reportReturnType)

And for `A.f`:

error: Cannot assign to attribute "x" for class "A[T@A]*"
    Type "tuple[T@A]" is not assignable to type "tuple[int] | tuple[str]"
      "tuple[T@A]" is not assignable to "tuple[int]"
        Tuple entry 1 is incorrect type
          Type "int* | str*" is not assignable to type "int"
            "str*" is not assignable to "int"
      "tuple[T@A]" is not assignable to "tuple[str]"
        Tuple entry 1 is incorrect type
          Type "int* | str*" is not assignable to type "str"
    ... (reportAttributeAccessIssue)

**Comment by erictraut** [comment]
Pyright is working as designed here, so I don't consider this a bug.

The typing spec has very little to say about value-constrained type variables. Different type checkers interpret them differently, and you'll find that there are many limitations based on the way they are interpreted. I recommend avoiding the use of value-constrained type variables. There's almost always a better approach that doesn't involve their use. To my knowledge, no other programming language or type system has such a concept — and for good reason. It's not clear that we will ever be able to reconcile and compose value-constrained type variables with other features of the Python type system. Unless/until someone attempts to clarify exactly how they should work in the typing spec, I don't plan to make any changes to pyright's current handling of them.

You can read more about value-constrained type variables in the [pyright documentation](https://microsoft.github.io/pyright/#/mypy-comparison?id=value-constrained-type-variables).
